tools/main.py

```diff
@@ -0,0 +1,131 @@
+import msgpack
+import csv
+import matplotlib.pyplot as plt
+import pandas as pd
+
+from Packet import *
+
+
+def retrieve_data(infile, boards, start=-1, end=float('inf')):
+    # Takes logfile and a list of boards' data to read
+    # Returns dictionary, 'data', with information for all boards using timestamp as keys
+    # Will only return data within specific timestamp if specified
+
+    data = dict()
+    previous_timestamp = 0
+    data[previous_timestamp] = {}
+
+    # Creates first item in data, at time=0 and values=None for each board
+    for board in boards:
+        data[previous_timestamp][board] = None
+
+    # Loop through log file to retrieve data, adds it to 'data'
+    # Turns each logged packet into a Packet object for simplicity.
+    infile.seek(0)
+    for raw_packet in msgpack.Unpacker(infile):
+        for packet in Packet.sort(raw_packet):
+
+            # Check if packet is within specified timerange
+            if packet.timestamp <= start or packet.timestamp >= end:
+                continue
+
+            # Check if packet is of a specified board
+            elif packet.name in boards:
+                # initialize the key
+                if packet.timestamp not in data:
+                    data[packet.timestamp] = {}
+
+                data[packet.timestamp][packet.name] = packet.value
+
+                # To prevent 'holes' in the dictionary, all other values are
+                # copied over from the previous timestamp
+                for board in boards:
+                    if board not in data[packet.timestamp]:
+                        data[packet.timestamp][board] = data[previous_timestamp][board]
+
+                previous_timestamp = packet.timestamp
+
+    # Remove the initial entry
+    del data[0]
+
+    data_list = list()
+    for timestamp, items in data.items():
+        data[timestamp]['Timestamp'] = timestamp
+        data_list.append(items)
+    del data
+
+    return data_list
+
+
+def to_csv(outfile, field_names, data):
+    # Turns dict from retrieve_data() into a csv compatible list
+    field_names.insert(0, 'Timestamp')
+
+    with open(outfile, 'w') as csvfile:
+        writer = csv.DictWriter(csvfile, fieldnames=field_names, lineterminator='\n')
+        writer.writeheader()
+        writer.writerows(data)
+
+
+def plot_data(infile):
+    df = pd.read_csv(infile)
+
+    ax = plt.gca()  # plot on same graph
+
+    for col in df.columns.values[1:]:
+        df.plot(x='Timestamp', y=col, kind='line', ax=ax)
+
+    plt.show()
+
+
+def main(infile):
+    # Read msgpack file
+    infile = open(infile, 'rb')
+
+    # Search log file to see which sensors and actuators have readings logged.
+    # Turns each logged packet into a Packet object for simplicity.
+    # Adds all found boards into 'available' set
+    available = set()
+    for raw_packet in msgpack.Unpacker(infile):
```

Comment:
we read the file twice (once here and once in retrieve data), it would be nice if we could wrap this all into one class and have the constructor read the file and some other methods get us the values we need when we need them